### PR TITLE
Use ea interface to calculate accumulator functions for datetimelike

### PR DESCRIPTION
**Pull request by phofl**
- [ ] closes #xxxx (Replace xxxx with the GitHub issue number)
- [ ] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [ ] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [ ] Added [type annotations](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#type-hints) to new arguments/methods/functions.
- [ ] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

cc @jbrockmendel Follow up from the other pr

As mentioned on the other pr, this changes the behavior (highlighted by the test).

### CODE REVIEW

File: pandas/core/array_algos/datetimelike_accumulations.py [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+    func : np.cumsum, np.cumprod, np.maximum.accumulate, np.minimum.accumulate
+    values : np.ndarray
+        Numpy array with the values (can be of any dtype that support the
+        operation).
+    skipna : bool, default True
+        Whether to skip NA.
+    """
+    try:
+        fill_value = {
+            np.cumprod: 1,
```
Comment (jbrockmendel) [inline pandas/core/array_algos/datetimelike_accumulations.py:36]: is np.cumprod relevant?

File: pandas/core/array_algos/datetimelike_accumulations.py [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+    skipna: bool = True,
+):
+    """
+    Accumulations for 1D datetimelike arrays.
+
+    Parameters
+    ----------
+    func : np.cumsum, np.cumprod, np.maximum.accumulate, np.minimum.accumulate
+    values : np.ndarray
+        Numpy array with the values (can be of any dtype that support the
```
Comment (jbrockmendel) [inline pandas/core/array_algos/datetimelike_accumulations.py:29]: note that these _will_ be modified

File: pandas/core/array_algos/datetimelike_accumulations.py [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+
+    result = func(y)
+    result[mask] = iNaT
+
+    if values.dtype.kind in ["m", "M"]:
+        return result.view(values.dtype.base)
+    return result
+
+
+def cumsum(values: np.ndarray, *, skipna: bool = True):
```
Comment (jbrockmendel) [inline pandas/core/array_algos/datetimelike_accumulations.py:59]: `-> np.ndarray`?

File: pandas/core/arrays/datetimelike.py [hunk 4]
```diff
@@ -1380,25 +1381,15 @@ def _addsub_object_array(self, other: np.ndarray, op):
...
-            # "_simple_new" of "NDArrayBacked"  [call-arg]
-            return type(self)._simple_new(
-                result, freq=self.freq, dtype=self.dtype  # type: ignore[call-arg]
-            )
+        op = getattr(datetimelike_accumulations, name)
+        result = op(self.copy(), skipna=skipna, **kwargs)
 
-        raise TypeError(f"Accumulation {name} not supported for {type(self)}")
+        return type(self)._simple_new(
+            result, freq=self.freq, dtype=self.dtype  # type: ignore[call-arg]
```
Comment (jbrockmendel) [inline pandas/core/arrays/datetimelike.py:1391]: i dont think retaining self.freq is going to be correct in general

File: pandas/tests/series/test_cumulative.py [hunk 5]
```diff
@@ -91,6 +91,26 @@ def test_cummin_cummax_datetimelike(self, ts, method, skipna, exp_tdi):
...
         tm.assert_series_equal(expected, result)
 
+    @pytest.mark.parametrize(
+        "func, exp",
+        [
+            ("cummin", pd.Period("2012-1-1", freq="D")),
+            ("cummax", pd.Period("2012-1-2", freq="D")),
+        ],
+    )
+    def test_cummin_cummax_period(self, func, exp):
```
Comment (jbrockmendel) [inline pandas/tests/series/test_cumulative.py:101]: thanks for checking this

File: pandas/core/array_algos/datetimelike_accumulations.py (hunk 1 above)
Comment (phofl) [reply pandas/core/array_algos/datetimelike_accumulations.py:36]: Hm I guess we can remove it

File: pandas/core/array_algos/datetimelike_accumulations.py (hunk 2 above)
Comment (phofl) [reply pandas/core/array_algos/datetimelike_accumulations.py:29]: Added

File: pandas/core/array_algos/datetimelike_accumulations.py (hunk 3 above)
Comment (phofl) [reply pandas/core/array_algos/datetimelike_accumulations.py:59]: Yep good point, this one works. The others aren't that easy...

File: pandas/core/arrays/datetimelike.py (hunk 4 above)
Comment (phofl) [reply pandas/core/arrays/datetimelike.py:1391]: Can you elaborate when we should set freq to None?

Comment (jbrockmendel) [reply pandas/core/arrays/datetimelike.py:1391]: Pretty much anything that isn't a no-op should be not-freq-preserving.  e.g. `pd.date_range("2016-01-01", periods=3)._data.cummin()`

Comment (phofl) [reply pandas/core/arrays/datetimelike.py]: Thx, makes sense. Should we retain the freq in cases where we only have one element or not retain in general?

Not easy to hit btw :) But added a test now

Comment (jbrockmendel) [reply pandas/core/arrays/datetimelike.py]: id just not-retain in general

Comment (phofl) [reply pandas/core/arrays/datetimelike.py]: Sounds good, that's how I implemented it right now

File: pandas/core/array_algos/datetimelike_accumulations.py [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+    values: np.ndarray,
+    *,
+    skipna: bool = True,
+):
+    """
+    Accumulations for 1D datetimelike arrays.
+
+    Parameters
+    ----------
+    func : np.cumsum, np.cumprod, np.maximum.accumulate, np.minimum.accumulate
```
Comment (mroeschke) [inline pandas/core/array_algos/datetimelike_accumulations.py:27]:
```suggestion
    func : np.cumsum, np.maximum.accumulate, np.minimum.accumulate
```

File: pandas/core/array_algos/datetimelike_accumulations.py [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+    values : np.ndarray
+        Numpy array with the values (can be of any dtype that support the
+        operation). Values is changed is modified inplace.
+    skipna : bool, default True
+        Whether to skip NA.
+    """
+    try:
+        fill_value = {
+            np.maximum.accumulate: np.iinfo(np.int64).min,
+            np.cumsum: 0,
```
Comment (mroeschke) [inline pandas/core/array_algos/datetimelike_accumulations.py:37]: Do we have tests for cumsum? Also I guess this only makes sense for timedelta as well?

Comment (phofl) [reply pandas/core/array_algos/datetimelike_accumulations.py:37]: Good point, no we don't. Added tests

**Review by mroeschke — APPROVED** [review]
LGTM merge when ready @phofl or @jbrockmendel

File: pandas/core/nanops.py [hunk 8]
```diff
@@ -1712,52 +1712,7 @@ def na_accum_func(values: ArrayLike, accum_func, *, skipna: bool) -> ArrayLike:
...
-            # For DatetimeTZDtype, view result as M8[ns]
-            npdtype = orig_dtype if isinstance(orig_dtype, np.dtype) else "M8[ns]"
-            # Item "type" of "Union[Type[ExtensionArray], Type[ndarray[Any, Any]]]"
-            # has no attribute "_simple_new"
-            result = type(values)._simple_new(  # type: ignore[union-attr]
-                result.view(npdtype), dtype=orig_dtype
-            )
-
-    elif skipna and not issubclass(values.dtype.type, (np.integer, np.bool_)):
+    if skipna and not issubclass(values.dtype.type, (np.integer, np.bool_)):
```
Comment (jbrockmendel) [inline pandas/core/nanops.py:1715]: maybe a comment/check that "mM" cases should not get here?

Comment (jbrockmendel) [reply pandas/core/nanops.py:1715]: (OK for follow-up, i can add this into my next "assorted" branch)

File: pandas/tests/arrays/datetimes/test_accumulator.py [hunk 9]
```diff
@@ -0,0 +1,46 @@
+import pytest
```
Comment (jbrockmendel) [inline pandas/tests/arrays/datetimes/test_accumulator.py:1]: can you call this test_cumulative.py to match tests/series/ and tests/frame/

Comment (phofl) [reply pandas/tests/arrays/datetimes/test_accumulator.py:1]: Done

File: pandas/core/nanops.py (hunk 8 above)
Comment (phofl) [reply pandas/core/nanops.py:1715]: Added an assert

**Review by mroeschke — APPROVED** [review]

**Comment by mroeschke** [comment]
Thanks @phofl

**Pull request merged**